File: 2021/day03/day03.swift
#!/usr/bin/env swift

import Foundation

public extension String {
	func character(at: Int) -> Character {
		self[index(startIndex, offsetBy: at)]
	}
}

// Advent of Code 2021
// Day 03

let content = try! String(contentsOfFile: "input.txt", encoding: .utf8)
let lines = content.components(separatedBy: "\n").filter { !$0.isEmpty }

let bitCount = lines[0].count

// Part One:
// What is the power consumption of the submarine?

var result: [Int] = .init(repeating: 0, count: bitCount)
for bit in 0..<bitCount {
    var count0 = 0, count1 = 0
    for line in 0..<lines.count {
        count0 += lines[line].character(at: bit) == "0" ? 1 : 0
        count1 += lines[line].character(at: bit) == "1" ? 1 : 0
    }

    result[bit] = count0 >= count1 ? 0 : 1
}

let gamma = Int(result.map(\.description).joined(), radix: 2)!
let epsilon = ~gamma & (Int(pow(2.0, Double(bitCount)) - 1))

print("Part One:", gamma * epsilon)

// Parto Two
// What is the life support rating of the submarine?

func shrubb(lines: [String], bit: Int, way: Bool) -> [String] {
    var count0 = 0, count1 = 0
    for line in 0 ..< lines.count {
        count0 += lines[line].character(at: bit) == "0" ? 1 : 0
        count1 += lines[line].character(at: bit) == "1" ? 1 : 0
    }

    let matchChar: Character = way ?
        (count0 > count1 ? "0" : "1") :
        (count0 <= count1 ? "0" : "1")

    return lines.filter { $0.character(at: bit) == matchChar }
}

func compute(_ lines: [String], _ way: Bool) -> Int {
    var temp = lines
    var bitPos = 0
    while temp.count > 1 {
        temp = shrubb(lines: temp, bit: bitPos, way: way)
        bitPos += 1
    }
    return Int(temp[0], radix: 2)!
}

let oxygen = compute(lines, true)
let co2 = compute(lines, false)

print("Part Two:", oxygen * co2)

File: 2021/day03/day03.py
#!/usr/bin/python3

# # Advent of Code 2021
# Day 03

if __name__ == '__main__':

    file = open('input.txt')
    lines = [line.strip() for line in file]

    midCount = len(lines) // 2
    bitCount = len(lines[0])

    # Part One:
    # What is the power consumption of the submarine?

    count1 = [0] * bitCount
    for line in lines:
         for i, c in enumerate(line):
            if c == '1':
                count1[i] += 1

    gamma = 0
    pow = 2 ** (bitCount - 1)
    for i in range(bitCount):
        if count1[i] >= midCount:
            gamma += pow
        pow //= 2

    epsilon = ~gamma & 2 ** bitCount - 1

    print( "Part One: {:d}".format(gamma * epsilon) )

    # Part Two:
    # What is the life support rating of the submarine?

    i = 0
    numbers = set(lines)
    while len(numbers) > 1:
        count0 = 0
        count1 = 0

        for n in numbers:
            if n[i] == "0":
                count0 += 1
            elif n[i] == "1":
                count1 += 1

        for n in tuple(numbers):
            if n[i] == ("1" if count1 < count0 else "0"):
                numbers.remove(n)

        i += 1

    o = int(next(iter(numbers)), 2)

    i = 0
    numbers = set(lines)
    while len(numbers) > 1:
        count0 = 0
        count1 = 0

        for n in numbers:
            if n[i] == "0":
                count0 += 1
            elif n[i] == "1":
                count1 += 1

        for n in tuple(numbers):
            if n[i] == ("0" if count1 < count0 else "1"):
                numbers.remove(n)

        i += 1

    c = int(next(iter(numbers)), 2)
    print( "Part Two: {:d}".format(o * c) )
